proof/refine/ARM/Schedule_R.thy

```diff
@@ -4210,6 +4209,1003 @@ lemma refillUnblockCheck_corres':
+                         readCurSc_def obj_at'_def ogets_def
+                  split: option.splits if_split_asm)
+  done
+
+lemma head_time_buffer_equiv:
+  "\<lbrakk>(s, s') \<in> state_relation; pspace_aligned s; pspace_distinct s; valid_objs s;
+    pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) (cur_sc s);
+    valid_refills' (ksCurSc s') s'; usage = usage'\<rbrakk>
+   \<Longrightarrow> head_time_buffer usage s = headTimeBuffer usage' s'"
+  apply (prop_tac "sc_at (cur_sc s) s")
+   apply (fastforce dest: valid_objs_valid_sched_context_size
+                    simp: vs_all_heap_simps obj_at_kh_kheap_simps is_sc_obj_def)
+  apply (frule state_relation_pspace_relation)
+  apply (frule sc_at_cross; simp?)
+  apply (frule state_relation_sc_relation; simp?)
+  apply (subst head_time_buffer_simp; simp?)
+  apply (subst headTimeBuffer_simp)
+   apply (clarsimp simp: state_relation_def)
+  apply (frule refill_hd_relation)
+    apply (clarsimp simp: valid_refills'_def obj_at_simps state_relation_def)
+   apply (clarsimp simp: sc_ko_at_valid_objs_valid_sc' opt_map_def obj_at_simps)
+  apply (clarsimp simp: obj_at_def sc_at_ppred_def obj_at'_def projectKOs state_relation_def
+                        maxReleaseTime_equiv opt_map_def vs_all_heap_simps refill_map_def)
+  done
+
+lemma refillSingle_sp:
+  "\<lbrace>P\<rbrace>
+   refillSingle scp
+   \<lbrace>\<lambda>rv s. P s \<and> (\<forall>ko. ko_at' ko scp s \<longrightarrow> rv = (scRefillHead ko = refillTailIndex ko))\<rbrace>"
+  apply (clarsimp simp: refillSingle_def)
+  apply wpsimp
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+crunches updateRefillHd, scheduleUsed
+   for valid_sched_context'[wp]: "valid_sched_context' sc"
+
+lemma handleOverrunLoopBody_corres:
+  "r = r'
+   \<Longrightarrow> corres (=) (\<lambda>s. sc_at (cur_sc s) s \<and> is_active_sc2 (cur_sc s) s
+                       \<and> pspace_aligned s \<and> pspace_distinct s
+                       \<and> pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) (cur_sc s))
+                  (\<lambda>s'. valid_refills' (ksCurSc s') s')
+                  (handle_overrun_loop_body r)
+                  (handleOverrunLoopBody r')"
+  apply (rule_tac Q="\<lambda>s'. sc_at' (ksCurSc s') s'" in corres_cross_add_guard)
+   apply (fastforce intro: sc_at_cross simp: state_relation_def)
+  apply (rule_tac Q="\<lambda>s'. is_active_sc' (ksCurSc s') s'" in corres_cross_add_guard)
+   apply (fastforce intro: is_active_sc'_cross simp: state_relation_def)
+  apply (clarsimp simp: handle_overrun_loop_body_def handleOverrunLoopBody_def)
+  apply (rule corres_split'[rotated 2, OF gets_sp getCurSc_sp])
+   apply (corressimp corres: getCurSc_corres)
+  apply (rule corres_split'[rotated 2, OF refill_single_sp refillSingle_sp])
+   apply (corressimp corres: refillSingle_corres)
+   apply (fastforce dest!: valid_objs'_valid_refills'
```

Comment:
I am guessing that these `valid_objs’_valid_refills’` (here and twice below) are probably not needed? If they are needed, I wonder where `valid_objs’` gets resurrected.

Reply:
Removed, including an instance in `scheduleUsed_corres`

Reply:
👍 I must have missed that one. While you are there, can you check one more thing? In `scheduleUsed_corres`, I think crossing of `valid_refills’` is now necessary and should be removed.

Reply:
I've removed that crossing, too